src/NovelRT/Graphics/GraphicsResourceManager.cpp

```diff
@@ -86,86 +139,138 @@ namespace NovelRT::Graphics
-        size_t sizeToAllocate = _graphicsDevice->GetMemoryAllocator()->GetSettings().MinimumBlockSize;
+        std::shared_ptr<GraphicsBuffer> newBuffer = nullptr;
 
-        while (sizeToAllocate < allocationSize)
+        if (bufferKind == GraphicsBufferKind::Constant)
         {
-            sizeToAllocate += minimumBlockSize;
+            newBuffer = _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
+                GraphicsBufferKind::Constant, GraphicsResourceAccess::Write, GraphicsResourceAccess::Read,
+                ((allocationSize > _tenMegabytesAsBytes) ? allocationSize : _tenMegabytesAsBytes));
         }
-
-        auto newBuffer = _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
-            GraphicsBufferKind::Vertex, GraphicsResourceAccess::None, GraphicsResourceAccess::Write,
-            ((allocationSize > _tenMegabytesAsBytes) ? allocationSize : _tenMegabytesAsBytes));
-
-        _vertexBuffers.emplace_back(newBuffer);
-
-        return newBuffer;
-    }
-
-    std::shared_ptr<GraphicsBuffer> GraphicsResourceManager::GetOrCreateConstantBufferForAllocationSize(
-        size_t allocationSize)
-    {
-        for (auto&& buffer : _constantBuffers)
+        else
         {
-            if (buffer->GetLargestFreeRegionSize() >= allocationSize)
+            size_t minimumBlockSize = _graphicsDevice->GetMemoryAllocator()->GetSettings().MinimumBlockSize;
+            size_t sizeToAllocate = _graphicsDevice->GetMemoryAllocator()->GetSettings().MinimumBlockSize;
+            while (sizeToAllocate < allocationSize)
             {
-                return buffer;
+                sizeToAllocate += minimumBlockSize;
             }
-        }
 
-        auto newBuffer = _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
-            GraphicsBufferKind::Constant, GraphicsResourceAccess::Write, GraphicsResourceAccess::Read,
-            ((allocationSize > _tenMegabytesAsBytes) ? allocationSize : _tenMegabytesAsBytes));
+            newBuffer = _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
+                bufferKind, GraphicsResourceAccess::None, GraphicsResourceAccess::Write,
+                ((allocationSize > _tenMegabytesAsBytes) ? allocationSize : _tenMegabytesAsBytes));
+        }
 
-        _constantBuffers.emplace_back(newBuffer);
+        switch (bufferKind)
+        {
+            case GraphicsBufferKind::Default:
+                _defaultBuffers.emplace_back(newBuffer);
+                break;
+            case GraphicsBufferKind::Vertex:
+                _vertexBuffers.emplace_back(newBuffer);
+                break;
+            case GraphicsBufferKind::Index:
+                _indexBuffers.emplace_back(newBuffer);
+                break;
+            case GraphicsBufferKind::Constant:
+                _constantBuffers.emplace_back(newBuffer);
+                break;
+        }
 
         return newBuffer;
```

Comment:
If the switch was changed to return the value of `emplace_back`, this should instead be changed to fail-fast/throw and return nullptr.